File: python3/Sort and Search/sort_array_quicksort_optimized.py
import random
from typing import List


def sort_array(nums: List[int]) -> List[int]:
    quicksort_optimized(nums, 0, len(nums) - 1)
    return nums

def quicksort_optimized(nums: List[int], left: int, right: int) -> None:
    if left >= right:
        return
    # Choose a pivot at a random index.
    random_index = random.randint(left, right)
    # Swap the randomly chosen pivot with the rightmost element to 
    # position the pivot at the rightmost index.
    nums[random_index], nums[right] = nums[right], nums[random_index]
    pivot_index = partition(nums, left, right)
    quicksort_optimized(nums, left, pivot_index - 1)
    quicksort_optimized(nums, pivot_index + 1, right)

def partition(nums: List[int], left: int, right: int) -> int:
    pivot = nums[right]
    lo = left
    # Move all numbers less than the pivot to the left, which
    # consequently positions all numbers greater than or equal to the
    # pivot to the right.
    for i in range(left, right):
        if nums[i] < pivot:
            nums[lo], nums[i] = nums[i], nums[lo]
            lo += 1
    # After partitioning, 'lo' will be positioned where the pivot should
    # be. So, swap the pivot number with the number at the 'lo' pointer.
    nums[lo], nums[right] = nums[right], nums[lo]
    return lo

File: java/Sort and Search/SortArrayQuicksortOptimized.java
public class SortArrayQuicksortOptimized {
    public int[] sortArray(int[] nums) {
        quicksortOptimized(nums, 0, nums.length - 1);
        return nums;
    }

    private void quicksortOptimized(int[] nums, int left, int right) {
        if (left >= right) return;
        // Choose a pivot at a random index.
        // Math.random() generate a double in [0.0, 1.0)
        // Scale the range from [0.0, 1.0) to [0, right - left + 1)
        int randomIndex = left + (int)(Math.random() * (right - left + 1));
        // Swap the randomly chosen pivot with the rightmost element to 
        // position the pivot at the rightmost index.
        swap(nums, randomIndex, right);
        int pivotIndex = partition(nums, left, right);
        quicksortOptimized(nums, left, pivotIndex - 1);
        quicksortOptimized(nums, pivotIndex + 1, right);
    }

    private int partition(int[] nums, int left, int right) {
        int pivot = nums[right];
        int lo = left;
        // Move all numbers less than the pivot to the left, which
        // consequently positions all numbers greater than or equal to the
        // pivot to the right.
        for (int i = left; i < right; i++) {
            if (nums[i] < pivot) {
                swap(nums, lo, i);
                lo++;
            }
        }
        // After partitioning, 'lo' will be positioned where the pivot should
        // be. So, swap the pivot number with the number at the 'lo' pointer.
        swap(nums, lo, right);
        return lo;
    }

    private void swap(int[] nums, int a, int b) {
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
}
